enhancements/installer/aws-ingress-nlb-security-group.md

```diff
@@ -0,0 +1,498 @@
+    - Configure Ingress rules in the Security Group to allow traffic on the ports defined in the Service's `spec.ports`. The source for these rules will be determined by the `service.beta.kubernetes.io/load-balancer-source-ranges` annotation on the Service (if present, otherwise default to allowing from all IPs).
+    - Configure Egress rules in the Security Group to allow traffic to the backend pods on the targetPort specified in the Service's `spec.ports` and the health check port. Initially, this should be restricted to the cluster's VPC CIDR or the specific CIDRs of the worker nodes.
+    - When creating the NLB using the AWS ELBv2 API, the CCM will include the ID of the newly created Security Group in the `SecurityGroups` parameter of the `CreateLoadBalancerInput.`
+  - When the Service is deleted, the CCM will also delete the associated Security Group, ensuring proper cleanup.
+
+#### OpenShift Managed (TBD)
+
+- ROSA Classic:
+    - TBD: Ensure the `install-config.yaml` option correctly writes the CIO manifests enabling NLB with managed Security Groups. This might involve changes in the Hive component responsible for cluster provisioning.
+
+- ROSA HCP:
+    - TBD: Investigate how Hypershift sets the Service Annotation of Security Group management when launching CIO. The `install-config.yaml` option should influence the Hosted Control Plane's CIO configuration. Need to validate if ROSA HCP follows the pattern of Hypershift or require additional configuration.
+
+### API Extensions
+
+> WIP/TBReviewed
+
+#### AWS Cloud Controller Manager (CCM)
+
+- Introduction of a new annotation constant: `ServiceAnnotationLoadBalancerManagedSecurityGroup = "service.beta.kubernetes.io/aws-load-balancer-managed-security-group"` in `pkg/providers/v1/aws.go`.
+
+```go
+// ServiceAnnotationLoadBalancerManagedSecurityGroup is the annotation used
+// on the service to specify the instruct CCM to manage the security group when creating a Network Load Balancer. When enabled,
+// the CCM creates the security group and it's rules. This option can not be used with annotations
+// "service.beta.kubernetes.io/aws-load-balancer-security-groups" and "service.beta.kubernetes.io/aws-load-balancer-extra-security-groups".
+const ServiceAnnotationLoadBalancerManagedSecurityGroup = "service.beta.kubernetes.io/aws-load-balancer-managed-security-group"
+```
+
+- Logic in the service controller within the CCM (`pkg/providers/v1/aws.go` and `pkg/providers/v1/aws_loadbalancer.go` ) to recognize and handle the new annotation when the service type is `NLB` (`ServiceAnnotationLoadBalancerType = "service.beta.kubernetes.io/aws-load-balancer-type"` ).
+- Functionality within the CCM to create and manage the lifecycle of AWS Security Groups for NLBs, including creating ingress and egress rules based on the service specification. This would likely involve using the AWS SDK for Go to interact with the EC2 API for creating and managing security groups.
+
```

Comment:
i wonder if the upstream changes to CCM will need to be behind a feature gate?

given that this will be a new feature to the CCM that the ALBO already controls, i'm guessing that we will need a way to ensure that users who run CCM and ALBO in the same cluster can control the behavior. perhaps a flag to the CCM.

Reply:
Good catch, Mike, thanks! That's make sense. I will search how it does in CCM and document here.

Reply:
[Cross-ref][xref] to the thread that we could enhance the config to improve the EP goal.

[xref]: https://github.com/openshift/enhancements/pull/1802/files#r2111532244